k8s-scheduler/src/main/java/com/vmware/dcm/AutoScope.java

```diff
@@ -19,18 +21,134 @@
+
+    SortHelper(final String tableName, final String groupField, final String sortField) {
+        this.tableName = tableName.toUpperCase();
+        this.groupField = groupField.toUpperCase();
+        this.sortField = sortField.toUpperCase();
+    }
+
+    public boolean isSortTable(final Record r) {
+        final String field = r.field(0).toString().toUpperCase();
+        return field.contains(this.tableName);
+    }
+
+    private int getSortVal(final Record r) {
+        return Integer.parseInt(r.get(this.sortField).toString());
+    }
+
+    public String getGroup(final Record r) {
+        return r.get(this.groupField).toString();
+    }
+
+    public int removeFromTopk(final List<Record> records, final Record toDel) {
+        int index = -1;
+        for (int i = 0; i < records.size(); i ++) {
+            final Record record = records.get(i);
+            if (record.equals(toDel)) {
+                index = i;
+                break;
+            }
+        }
+        return index;
+    }
+
+    public int addToTopk(final List<Record> records, final Record toAdd) {
+        int minVal = getSortVal(toAdd);
+        int index = -1;
+        for (int i = 1; i < records.size(); i += 1) {
+            final int val = getSortVal(records.get(i));
+            if (val < minVal) {
+                index = i;
+                minVal = val;
+            }
+        }
+        return index;
+    }
+}
+
+class TopkPerGroup implements DeltaCallBack {
+    private static final Logger LOG = LoggerFactory.getLogger(TopkPerGroup.class);
+
+    private final int limit;
+    private SortHelper helper;
+    private Map<String, List<Record>> topk;
+
+    public TopkPerGroup(final String tableName, final String groupField, final String sortField, final int limit) {
+        this.limit = limit;
+        this.topk = new HashMap<>();
+        this.helper = new SortHelper(tableName, groupField, sortField);
+    }
+
+    public List<Record> getAll() {
+        final List<Record> results = new ArrayList<>();
+        for (final Map.Entry<String, List<Record>> entry: topk.entrySet()) {
+            results.addAll(entry.getValue());
+        }
+        return results;
+    }
+
+    @Override
+    public void processDelta(final DeltaType type, final Record r) {
+        if (helper.isSortTable(r)) {
+            final String group = helper.getGroup(r);
+            if (type == DeltaCallBack.DeltaType.ADD) {
+                final List<Record> records = topk.getOrDefault(group, new ArrayList<>());
+                if (records.size() < this.limit) {
+                    records.add(r);
+                } else {
+                    final int index = helper.addToTopk(records, r);
+                    if (index >= 0) {
+                        records.set(index, r);
+                    }
+                }
+                topk.put(group, records);
+            } else {
+                final List<Record> records = topk.getOrDefault(group, new ArrayList<>());
```

Comment:
The logic needs to replace existing elements from the full list when deletions happen.